cdap-app-fabric/src/main/java/co/cask/cdap/app/services/AbstractServiceDiscoverer.java

```diff
@@ -85,4 +101,85 @@ private URL createURL(@Nullable Discoverable discoverable, String applicationId,
+    Discoverable discoverable = new RandomEndpointStrategy(discovered).pick(1, TimeUnit.SECONDS);
+    if (discoverable != null) {
+      InetSocketAddress address = discoverable.getSocketAddress();
+      String path = String.format("http://%s:%d%s/namespaces/%s/streams/%s", address.getHostName(), address.getPort(),
+                                  Constants.Gateway.API_VERSION_3, namespaceId, stream);
+      if (batch) {
+        path = String.format("%s/batch", path);
+      }
+      try {
+        return new URL(path);
+      } catch (MalformedURLException e) {
+        LOG.error("Got exception while creating StreamURL", e);
+      }
+    }
+    throw new StreamWriteException("Stream Service Endpoint not found");
+  }
+
+  private void writeToStream(String stream, HttpRequest request) throws IOException, StreamWriteException {
+    HttpResponse response = HttpRequests.execute(request);
+    if (response.getResponseCode() == HttpResponseStatus.NOT_FOUND.code()) {
+      throw new StreamWriteException(String.format("Stream %s not found", stream));
+    }
+  }
+
+  @Override
+  public void write(String stream, String data) throws IOException, StreamWriteException {
+    write(stream, data, Maps.<String, String>newHashMap());
+  }
+
+  @Override
+  public void write(String stream, String data, Map<String, String> headers) throws IOException, StreamWriteException {
+    URL streamURL = getStreamURL(stream);
+    HttpRequest request = HttpRequest.post(streamURL).withBody(data).addHeaders(headers).build();
+    writeToStream(stream, request);
+  }
+
+  @Override
+  public void write(String stream, ByteBuffer data) throws IOException, StreamWriteException {
+    write(stream, data, Maps.<String, String>newHashMap());
+  }
+
+  @Override
+  public void write(String stream, ByteBuffer data, Map<String, String> headers) throws IOException,
+    StreamWriteException {
+    URL streamURL = getStreamURL(stream);
+    HttpRequest request = HttpRequest.post(streamURL).withBody(data).addHeaders(headers).build();
+    writeToStream(stream, request);
+  }
+
+  @Override
+  public void write(String stream, StreamEventData data) throws IOException, StreamWriteException {
+    write(stream, data.getBody(), data.getHeaders());
+  }
+
+  @Override
+  public void writeInBatch(String stream, File file, String contentType) throws IOException, StreamWriteException {
+    URL url = getStreamURL(stream, true);
+    HttpRequest request = HttpRequest.post(url).withBody(file).addHeader(HttpHeaders.CONTENT_TYPE, contentType).build();
+    writeToStream(stream, request);
+  }
+
+  @Override
+  public StreamBatchWriter writeInBatch(String stream, String contentType) throws IOException, StreamWriteException {
+    URL url = getStreamURL(stream, true);
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+    connection.setRequestMethod(HttpMethod.POST.name());
+    connection.setReadTimeout(15000);
+    connection.setConnectTimeout(15000);
```

Comment:
So the user can use the `StreamBatchWriter` returned only for this time, right? We should document that.